Add client cert/key to command line args and options

pims opened the pull request:
Hi there 👋 ,

Thanks for creating Hurl and open sourcing it, it's really neat!
I noticed that to fully support mTLS, the `--cert` and `--key` cli flags were not yet supported.

This PR adds support for configuring the client cert and private key for establishing a mutual TLS connection to a server.

Example:
```
GET https://localhost:8443/
[Options]
cacert: out/HurlCA.crt
client_cert: out/client.crt
client_key: out/client.key

HTTP 200
```

outputs:
```
Request can be run with the following curl command:
* curl 'https://localhost:8443/' --cacert out/HurlCA.crt --cert out/client.crt --key out/client.key
```

This is my first Rust PR, so I probably got a few things wrong. There's a decent amount of duplication that could probably be simplified.

Naming wise, I've used `client_cert` and `client_key` because `cert/key` felt a little too generic, but happy to stick to the curl naming if you think it's better.

Let me know if this is a feature you think makes sense to be added and your feedback is obviously welcome.
Merci!

fabricereix commented:
Thanks @pims for your PR.
That's something we wanted to have (#166).

Relating to the naming, I would tend to stick to the same name used in curl.
It will also be great to add an integration test related to these new options in `/integration/ssl`

To merge into master, you will also need to sign your commits (https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits)

jcamiel commented on packages/hurl/src/cli/options.rs:
```diff
@@ -34,6 +34,8 @@ use crate::runner::Value;
 #[derive(Clone, Debug, PartialEq, Eq)]
 pub struct CliOptions {
     pub cacert_file: Option<String>,
+    pub clientcert_file: Option<String>,
```
Maybe change to fields to `client_cert_file` and `client_key_file`?

jcamiel commented on packages/hurl/src/cli/options.rs:
```diff
@@ -105,6 +107,20 @@ pub fn app(version: &str) -> Command {
         )
+        .arg(
+            clap::Arg::new("clientcert_file")
+                .long("cert")
```
We can add here the short name for this option `.short('E')` to mirror curl options

jcamiel commented on packages/hurl/src/http/options.rs:
```diff
@@ -20,6 +20,8 @@ use std::time::Duration;
 #[derive(Debug, Clone)]
 pub struct ClientOptions {
     pub cacert_file: Option<String>,
+    pub clientcert_file: Option<String>,
```
Same remark as previous `clientcert_file` -> `client_cert_file`, `clientkey_file` -> `client_key_file`

jcamiel commented on packages/hurl/src/runner/entry.rs:
```diff
@@ -300,6 +302,14 @@ pub fn get_entry_options(
                     }
+                    EntryOption::ClientCert(option) => {
+                        runner_options.clientcert_file = Some(option.filename.value.clone());
+                        logger.debug(format!("client_cert: {}", option.filename.value).as_str());
```
We're going to change the option name (to be like culr) so the logs will become 

```rust
logger.debug(format!("cert: {}", option.filename.value).as_str());
...
logger.debug(format!("key: {}", option.filename.value).as_str());
```

jcamiel commented on packages/hurl/src/runner/runner_options.rs:
```diff
@@ -26,6 +26,8 @@ use std::time::Duration;
 #[derive(Clone, Debug, PartialEq, Eq)]
 pub struct RunnerOptions {
     pub cacert_file: Option<String>,
+    pub clientcert_file: Option<String>,
```
Same remark as previous clientcert_file -> client_cert_file, clientkey_file -> client_key_file

jcamiel commented on packages/hurl_core/src/format/html.rs:
```diff
@@ -299,6 +301,40 @@ impl Htmlable for CaCertificateOption {
+        add_line_terminators(&mut buffer, self.line_terminators.clone());
+        buffer.push_str("<span class=\"line\">");
+        buffer.push_str(self.space0.to_html().as_str());
+        buffer.push_str("<span class=\"string\">client_cert</span>");
```
`client_cert` => `cert`

jcamiel commented on packages/hurl_core/src/format/html.rs:
```diff
@@ -299,6 +301,40 @@ impl Htmlable for CaCertificateOption {
+        add_line_terminators(&mut buffer, self.line_terminators.clone());
+        buffer.push_str("<span class=\"line\">");
+        buffer.push_str(self.space0.to_html().as_str());
+        buffer.push_str("<span class=\"string\">client_key</span>");
```
`client_key` => `key`

jcamiel commented on packages/hurlfmt/src/format/token.rs:
```diff
@@ -845,6 +847,48 @@ impl Tokenizable for CaCertificateOption {
+                .collect(),
+        );
+        tokens.append(&mut self.space0.tokenize());
+        tokens.push(Token::String("client_cert".to_string()));
```
`client_cert` => `cert`

jcamiel commented on packages/hurlfmt/src/format/token.rs:
```diff
@@ -845,6 +847,48 @@ impl Tokenizable for CaCertificateOption {
+                .collect(),
+        );
+        tokens.append(&mut self.space0.tokenize());
+        tokens.push(Token::String("client_key".to_string()));
```
`client_key`=> `key`

jcamiel commented on packages/hurl_core/src/parser/sections.rs:
```diff
@@ -386,6 +388,50 @@ fn option_cacert(reader: &mut Reader) -> ParseResult<'static, EntryOption> {
+fn option_clientcert(reader: &mut Reader) -> ParseResult<'static, EntryOption> {
+    let line_terminators = optional_line_terminators(reader)?;
+    let space0 = zero_or_more_spaces(reader)?;
+    try_literal("client_cert", reader)?;
```
`client_cert` => `cert`

jcamiel commented on packages/hurl_core/src/parser/sections.rs:
```diff
@@ -386,6 +388,50 @@ fn option_cacert(reader: &mut Reader) -> ParseResult<'static, EntryOption> {
+fn option_clientkey(reader: &mut Reader) -> ParseResult<'static, EntryOption> {
+    let line_terminators = optional_line_terminators(reader)?;
+    let space0 = zero_or_more_spaces(reader)?;
+    try_literal("client_key", reader)?;
```
`client_key`=> `key`

pims commented:
Thanks @fabricereix and @jcamiel for the review.
I’ll push changes as new commits and once it looks good squash and sign the commit, unless you’d prefer otherwise.

jcamiel commented:
Thanks for your work @pims. As stated in [CONTRIBUTING.md](https://github.com/Orange-OpenSource/hurl/blob/master/CONTRIBUTING.md), you'll have to rebase also your PR to appease our CI for merging. We do this do keek a linear Git history. Don't hesitate to ask anything if you've issues on your PR!

pims commented:
@jcamiel I believe I've addressed all your comments around naming.
@fabricereix I could use some advice on how to best add an integration test.

On the integration test topic, would it make sense to move the `Tests units and integration tests ` into a shell script so it can be used locally and in CI?

fabricereix commented:
@pims 
We can add an integration test once it is merged.
Could you remove the changes you have done in the existing integ test?
You also need to rebase and sign your commits.

jcamiel commented:
@pims if you need some helps to configure your env for commit signing (https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification) I can help you, don't hesitate.

pims commented:
@jcamiel thanks for the pointer. Turns out my new laptop wasn't properly configured for signing commits by default.
I've removed the integration changes, so hopefully this PR will be in a mergeable state. Thanks both for your patience .

fabricereix commented:
/accept

fabricereix commented:
sorry @pims, you have to rebase again

fabricereix commented:
/accept

fabricereix commented:
/accept --force

pims commented:
Thanks for the support getting this PR merged @fabricereix and @jcamiel.

How should we approach the integration tests?

jcamiel commented:
At minimum, we should document how to run the unit test and the integration tests in CONTRIBUTING.md so we can explain that we want integration tests on news features. Scripts are already here and can be launched locally, we should document them and explain also how to setup a local env.

jcamiel commented:
Fixes #1129 .
